File: src/main/java/com/yahoo/sdvornik/db/IndexSearcher.java
package com.yahoo.sdvornik.db;


public class IndexSearcher <T extends Comparable> {

  private T[] array;

  public IndexSearcher(T[] array) {
    this.array = array;
  }

  public int binarySearch(T key) {
    int low = 0;
    int high = array.length - 1;

    while (low <= high) {
      int mid = (low + high) >>> 1;
      int cmp = array[mid].compareTo(key);

      if (cmp < 0) low = mid + 1;
      else if (cmp > 0) high = mid - 1;
      else return mid;
    }
    if(low > array.length - 1) return array.length -1;
    return low-1;
  }
}

File: src/main/scala/com/yahoo/sdvornik/util/IndexSearcher.scala
package com.yahoo.sdvornik.util

class IndexSearcher[T : Ordering](array: Array[T]) {

  def binarySearch(key: T)(implicit ord: Ordering[T]): Int = {
    var low = 0
    var high = array.length - 1
    while ( {
      low <= high
    }) {
      val mid = (low + high) >>> 1
      val cmp = ord.compare(array(mid),key)
      if (cmp < 0) low = mid + 1
      else if (cmp > 0) high = mid - 1
      else return mid
    }
    if (low > array.length - 1) return array.length - 1
    low - 1
  }
}
